janitor/functions/conditional_join.py

```diff
@@ -1151,9 +695,421 @@ def _not_equal_indices(
+):
+    """Return indices if more conditions exist."""
+
+    for condition in conditions:
+        left_c, right_c, op = condition
+        left_c = extract_array(left_c, extract_numpy=True)[left_index]
+        right_c = extract_array(right_c, extract_numpy=True)[right_index]
+        op = operator_map[op]
+        mask = op(left_c, right_c)
+        if not mask.any():
+            return None
+        if is_extension_array_dtype(mask):
+            mask = mask.to_numpy(dtype=bool, na_value=False)
+        if not mask.all():
+            left_index = left_index[mask]
+            right_index = right_index[mask]
+
+    return left_index, right_index
+
+
+def _multiple_conditional_join_ne(
+    df: pd.DataFrame, right: pd.DataFrame, conditions: list
+) -> tuple:
+    """
+    Get indices for multiple conditions,
+    where all the operators are `!=`.
+
+    Returns a tuple of (left_index, right_index)
+    """
+
+    first, *rest = conditions
+    left_on, right_on, op = first
+
+    # get indices from the first condition
+    result = _generic_func_cond_join(
+        df[left_on], right[right_on], op, multiple_conditions=False
+    )
+
+    if result is None:
+        return None
+
+    rest = (
+        (df[left_on], right[right_on], op) for left_on, right_on, op in rest
+    )
+
+    return _generate_indices(*result, rest)
+
+
+def _multiple_conditional_join_eq(
+    df: pd.DataFrame, right: pd.DataFrame, conditions: list
+) -> tuple:
+    """
+    Get indices for multiple conditions,
+    if any of the conditions has an `==` operator.
+
+    Returns a tuple of (df_index, right_index)
+    """
+
+    eqs = [
+        (left_on, right_on)
+        for left_on, right_on, op in conditions
+        if op == _JoinOperator.STRICTLY_EQUAL.value
+    ]
+
+    left_on, right_on = zip(*eqs)
+    left_on = [*left_on]
+    right_on = [*right_on]
+
+    # get merge indices
+    left_index, right_index = _MergeOperation(
+        df, right, left_on=left_on, right_on=right_on, sort=False, copy=False
+    )._get_join_indexers()
+
+    if not left_index.size > 0:
+        return None
+
+    rest = (
+        (df[left_on], right[right_on], op)
+        for left_on, right_on, op in conditions
+        if op != _JoinOperator.STRICTLY_EQUAL.value
+    )
+
+    return _generate_indices(left_index, right_index, rest)
+
+
+def _multiple_conditional_join_le_lt(
+    df: pd.DataFrame, right: pd.DataFrame, conditions: list
+) -> tuple:
+    """
+    Get indices for multiple conditions,
+    where `>/>=` or `</<=` is present,
+    and there is no `==` operator.
+
+    Returns a tuple of (df_index, right_index)
+    """
+
+    le_lt = None
+    ge_gt = None
+    for condition in conditions:
+        *_, op = condition
+        if op in less_than_join_types:
+            le_lt = condition
+        elif op in greater_than_join_types:
+            ge_gt = condition
+        if le_lt and ge_gt:
+            break
```

Comment:
Could you help me understand why we're breaking here? Suppose if conditions is written like:
```python
('a', 'b', '<'),
('b', 'c', '>'),
('c', 'd', '<'),
```
Won't the third condition be included in `rest`?
Or would that be expected behaviour?

(Actually I think this is stemming from me not quite understanding the logic for `rest` and its place in `_generate_indices`. Would be grateful if you could summarize the logic here haha)

Reply:
added an explanation to the code; hopefully it clarifies things